File: server/src/lib/recall.ts
import type { Trail } from "./types";

/**
 * Matches an incoming question against trails the team has already recorded.
 *
 * Why lexical and not embeddings: recall sits directly in front of the model
 * call, so it has to be effectively free — an embedding round-trip would cost
 * most of the latency it is meant to save, and would need a vector store to
 * justify itself. Token overlap against the original question plus curated
 * aliases is crude, but it is instant, debuggable, and runs with no
 * infrastructure. The threshold is tuned to prefer a model call over a wrong
 * trail: showing someone a confidently irrelevant walkthrough is far more
 * damaging to trust than making them wait three seconds.
 */

const STOPWORDS = new Set([
  "a", "an", "the", "how", "do", "does", "did", "i", "you", "we", "to", "in", "on",
  "at", "of", "for", "is", "are", "was", "can", "could", "would", "should", "my",
  "me", "it", "this", "that", "with", "from", "and", "or", "but", "get", "got",
  "where", "what", "when", "why", "which", "please", "help", "need", "want",
  "there", "here", "so", "if", "then", "be", "am", "as", "by", "into", "out",
]);

function tokenize(input: string): string[] {
  return input
    .toLowerCase()
    .replace(/[^a-z0-9\s]/g, " ")
    .split(/\s+/)
    .filter((t) => t.length > 1 && !STOPWORDS.has(t));
}

/**
 * Weighted overlap of query tokens against a candidate string.
 * Returns 0..1 — the share of the query that the candidate accounts for.
 */
function overlap(queryTokens: string[], candidate: string): number {
  if (queryTokens.length === 0) return 0;
  const candidateTokens = new Set(tokenize(candidate));
  if (candidateTokens.size === 0) return 0;

  let hits = 0;
  for (const t of queryTokens) {
    if (candidateTokens.has(t)) {
      hits += 1;
      continue;
    }
    // Cheap stemming: catches export/exporting, variable/variables.
    for (const c of candidateTokens) {
      if (c.length > 4 && t.length > 4 && (c.startsWith(t) || t.startsWith(c))) {
        hits += 0.75;
        break;
      }
    }
  }
  return Math.min(1, hits / queryTokens.length);
}

export interface RecallHit {
  trail: Trail;
  score: number;
}

/**
 * Below this, we call the model instead. Set high on purpose — see the note
 * at the top of this file about the asymmetric cost of a wrong match.
 */
const THRESHOLD = 0.62;

/**
 * @param question The user's raw question.
 * @param trails   Everything the team has recorded.
 * @returns The best match above threshold, or null to fall through to the model.
 */
export function recall(question: string, trails: Trail[]): RecallHit | null {
  const q = tokenize(question);
  if (q.length === 0) return null;

  let best: RecallHit | null = null;

  for (const trail of trails) {
    // The original question is the strongest signal; aliases exist to catch
    // the phrasings that same question arrives in. Title and app are weaker
    // and only break ties.
    const score = Math.max(
      overlap(q, trail.question),
      ...trail.aliases.map((a) => overlap(q, a)),
      overlap(q, trail.title) * 0.9,
      overlap(q, trail.app) * 0.5,
    );
    if (!best || score > best.score) best = { trail, score };
  }

  return best && best.score >= THRESHOLD ? best : null;
}

/** Free-text filter for the library view. Looser than recall — the user is browsing, not asking. */
export function searchTrails(query: string, trails: Trail[]): Trail[] {
  const q = tokenize(query);
  if (q.length === 0) return trails;
  return trails
    .map((trail) => ({
      trail,
      score: Math.max(
        overlap(q, trail.title),
        overlap(q, trail.question),
        overlap(q, trail.app),
        ...trail.aliases.map((a) => overlap(q, a)),
      ),
    }))
    .filter((r) => r.score > 0.2)
    .sort((a, b) => b.score - a.score)
    .map((r) => r.trail);
}

File: server/recall.go
package main

import (
	"regexp"
	"strings"
)

// Matching an incoming question against trails the team has already recorded.
//
// Why lexical rather than embeddings: recall sits directly in front of the
// model call, so it has to be effectively free. An embedding round-trip would
// cost most of the latency it exists to save, and would need a vector store to
// justify itself. Token overlap against the original question plus curated
// aliases is crude, but it is instant, debuggable, and needs no infrastructure.
//
// The threshold is tuned to prefer a model call over a wrong trail. Showing
// someone a confidently irrelevant walkthrough damages trust far more than
// making them wait three seconds, so the costs are asymmetric and the threshold
// is too.

var stopwords = map[string]bool{
	"a": true, "an": true, "the": true, "how": true, "do": true, "does": true,
	"did": true, "i": true, "you": true, "we": true, "to": true, "in": true,
	"on": true, "at": true, "of": true, "for": true, "is": true, "are": true,
	"was": true, "can": true, "could": true, "would": true, "should": true,
	"my": true, "me": true, "it": true, "this": true, "that": true, "with": true,
	"from": true, "and": true, "or": true, "but": true, "get": true, "got": true,
	"where": true, "what": true, "when": true, "why": true, "which": true,
	"please": true, "help": true, "need": true, "want": true, "there": true,
	"here": true, "so": true, "if": true, "then": true, "be": true, "am": true,
	"as": true, "by": true, "into": true, "out": true,
}

var nonAlphanum = regexp.MustCompile(`[^a-z0-9\s]`)

func tokenize(input string) []string {
	cleaned := nonAlphanum.ReplaceAllString(strings.ToLower(input), " ")
	out := make([]string, 0, 8)
	for _, tok := range strings.Fields(cleaned) {
		if len(tok) > 1 && !stopwords[tok] {
			out = append(out, tok)
		}
	}
	return out
}

// overlap returns the share of the query that a candidate string accounts for,
// in [0,1].
func overlap(queryTokens []string, candidate string) float64 {
	if len(queryTokens) == 0 {
		return 0
	}
	candidateTokens := make(map[string]bool)
	for _, t := range tokenize(candidate) {
		candidateTokens[t] = true
	}
	if len(candidateTokens) == 0 {
		return 0
	}

	hits := 0.0
	for _, t := range queryTokens {
		if candidateTokens[t] {
			hits++
			continue
		}
		// Cheap stemming: catches export/exporting, variable/variables.
		for c := range candidateTokens {
			if len(c) > 4 && len(t) > 4 && (strings.HasPrefix(c, t) || strings.HasPrefix(t, c)) {
				hits += 0.75
				break
			}
		}
	}

	score := hits / float64(len(queryTokens))
	if score > 1 {
		return 1
	}
	return score
}

// RecallHit is a matched trail and the score that matched it.
type RecallHit struct {
	Trail Trail
	Score float64
}

// recallThreshold is set high on purpose — see the note at the top of this file.
const recallThreshold = 0.62

// recall returns the best match above threshold, or nil to fall through to the
// model.
func recall(question string, trails []Trail) *RecallHit {
	q := tokenize(question)
	if len(q) == 0 {
		return nil
	}

	var best *RecallHit
	for i := range trails {
		t := trails[i]

		// The original question is the strongest signal; aliases exist to catch
		// the phrasings that same question arrives in. Title and app are weaker
		// and only break ties.
		score := overlap(q, t.Question)
		for _, alias := range t.Aliases {
			if s := overlap(q, alias); s > score {
				score = s
			}
		}
		if s := overlap(q, t.Title) * 0.9; s > score {
			score = s
		}
		if s := overlap(q, t.App) * 0.5; s > score {
			score = s
		}

		if best == nil || score > best.Score {
			best = &RecallHit{Trail: t, Score: score}
		}
	}

	if best != nil && best.Score >= recallThreshold {
		return best
	}
	return nil
}

// searchTrails is the free-text filter for browsing. Looser than recall,
// because the user is looking rather than asking.
func searchTrails(query string, trails []Trail) []Trail {
	q := tokenize(query)
	if len(q) == 0 {
		return trails
	}

	type scored struct {
		trail Trail
		score float64
	}
	matches := make([]scored, 0, len(trails))

	for _, t := range trails {
		score := overlap(q, t.Title)
		for _, c := range append([]string{t.Question, t.App}, t.Aliases...) {
			if s := overlap(q, c); s > score {
				score = s
			}
		}
		if score > 0.2 {
			matches = append(matches, scored{t, score})
		}
	}

	// Simple insertion sort by score — the list is a team's trails, not a corpus.
	for i := 1; i < len(matches); i++ {
		for j := i; j > 0 && matches[j].score > matches[j-1].score; j-- {
			matches[j], matches[j-1] = matches[j-1], matches[j]
		}
	}

	out := make([]Trail, len(matches))
	for i, m := range matches {
		out[i] = m.trail
	}
	return out
}
